### DISCUSSION TITLE
is it possible to use ripgrep as a library inside of another Rust program?

### DISCUSSION THREAD

**Discussion by JordyScript** [BurntSushi/ripgrep#2067] [Q&A] [answered]
I want to ship ripgrep functionality as part of a larger Rust program (using Tauri), but do not want to rely on calling ripgrep as a separate shell process. How feasible is this? Is there any documentation that you can point to that could help with this?

**Comment by BurntSushi** [comment, marked as answer]
If you cannot create a child process, then your only option is to port ripgrep to Javascript or to devise a C API of ripgrep's various crates. Either is technically possible, but "feasible" is not a word I would use to describe it unless you have a lot of resources to throw at it.

I suggest shelling out, like VS Code does. Or writing your application in Rust and make use of ripgrep's crates as-is, and compile to WASM.

**Reply by JordyScript** [reply]
Thank you for your response, but I'm not sure if you've understood me correctly. So the way Tauri works is: you have a javascript frontend that communicates with a rust backend. My question is: can I use ripgrep as a Rust library from within the Rust backend instead of a shell process? In other words, how well does the source code of ripgrep lend itself to being used as a library to build other Rust programs with?

**Reply by BurntSushi** [reply]
You want https://docs.rs/grep

There is little high level documentation, and the APIs are lower level than the abstraction that ripgrep's CLI gives you. This is what I was referring to in the last sentence of my answer above.

**Reply by JordyScript** [reply]
Thank you very much !
